### PR TITLE
fixed compilation error with gcc version 7.1.1

### PR DESCRIPTION
**Pull request by damiel**
Pretty weird issue with gcc 7.1.1 on arch linux, took me a while to figure out.

The stuff regading dmap is especially weird, why the hell would we use stdlib print calls in there?

### CODE REVIEW

**Comment by BielBdeLuna** [comment]
I think this is due the dmap code being a port from the doom3 engine where things where handled in a older fashion way.

**Comment by damiel** [comment]
@BielBdeLuna Thanks for the explanation, this makes sense.

**Comment by DanielGibson** [comment]
> The stuff regading dmap is especially weird, why the hell would we use stdlib print calls in there?

it doesn't, and your fix probably breaks it.
It didn't call `printf()` which prints to stdout, but sprintf(), which puts the formatted string into the buffer given as first argument (`path` in this case). If you can tell us the compiler error message, maybe we can figure out a proper fix :)

I'm not sure about the idStrStatic fixes either - if they explicitly called `idStr::operator=()` (which should be ok, as idStrStatic is derived from idStr), they probably meant to do that. Again, the actual compiler would be helpful to figure out what goes wrong there.

File: neo/idlib/StrStatic.h [hunk 1]
```diff
@@ -102,39 +102,39 @@ class idStrStatic : public idStr
 	{
 		buffer[ 0 ] = '\0';
 		SetStaticBuffer( buffer, _size_ );
-		idStr::operator=( b );
+		idStrStatic::operator=( b );
```
Comment (DanielGibson) [inline neo/idlib/StrStatic.h]: maybe (probably) it was indented to call the baseclasses operator=()
(same of the other occurences)

File: neo/tools/compilers/dmap/dmap.cpp [hunk 2]
```diff
@@ -364,7 +364,7 @@ void Dmap( const idCmdArgs& args )
 	// if this isn't a regioned map, delete the last saved region map
 	if( passedName.Right( 4 ) != ".reg" )
 	{
-		std::sprintf( path, "%s.reg", dmapGlobals.mapFileBase );
+		common->Printf( ( path, "%s.reg", dmapGlobals.mapFileBase ) );
```
Comment (DanielGibson) [inline neo/tools/compilers/dmap/dmap.cpp]: the original call formats the path char array, I think that common->Printf() just prints to the console.

**Comment by damiel** [comment]
Here i hope this works: <script src="https://gist.github.com/damiel/a85ace2a9ab1228aea82ce2ccfe4c16c.js"></script>

**Comment by damiel** [comment]
Here are the error messages regarding idStr/idStrStatic: https://bpaste.net/show/8a19c7480bbf

And here the error messages regarding dmap: https://bpaste.net/show/f6b3abbcc533

File: neo/tools/compilers/dmap/dmap.cpp (hunk 2 above)
Comment (damiel) [inline neo/tools/compilers/dmap/dmap.cpp]: I see, i did some reading and i agree this is not the way to go.

**Comment by DanielGibson** [comment]
std::sprintf() should probably be replaced with variant of snprintf().
`idStr::snPrintf()` should work (note that it's a static method that operates on a char buffer and not on an idStr)

The `operator=()` thing is kinda fishy, as idStr only overloads operator= for `idStr` and `const char*`, not bool, float etc.
I *think* that it (before gcc7) uses `operator=(const idStr& text)` and implicitly casts bool, float etc to idStr with idStr's constructors for these types.
Long story short: Try replacing `idStr::operator=( b );` with `idStr::operator=( idStr( b ) );` (and similar for the other cases)

**Comment by damiel** [comment]
Okay, i adjusted the idStr stuff now with the solution you suggested, i'll figure out the rest regarding dmap.

**Comment by damiel** [comment]
Once the build is done and you are okay with this solution, may i squash the commits into one?

**Comment by DanielGibson** [comment]
this looks better; for the sprintf()-fixes you could even use sizeof(path) instead of hardcoding 1024, in case someone wants to change paths size later.

squashing the commits sounds like a good idea :)

**Comment by damiel** [comment]
Done, thanks alot for your feedback!

**Comment by shmerl** [comment]
Why not use `::snprintf` ?

**Comment by BielBdeLuna** [comment]
there shouldn't be any reference to doom classic in OTE

**Comment by DanielGibson** [comment]
> Why not use ::snprintf ?

it's not properly supported in MSVC

**Comment by DanielGibson** [comment]
I guess rbdoom3bfg could (should) use this fix, but might need some more work for doom classic

**Comment by shmerl** [comment]
@DanielGibson: There actually already exists a [merge request with such fixes](https://github.com/RobertBeckebans/RBDOOM-3-BFG/pull/372).

**Comment by shmerl** [comment]
It still happens on the master by the way, for example:
```
[ 65%] Building CXX object neo/CMakeFiles/OpenTechEngine.dir/d3xp/MultiplayerGame.cpp.o
.../build/OpenTechBFG/neo/d3xp/MultiplayerGame.cpp: In member function ‘const char* BFG::idMultiplayerGame::GameTime()’:
.../build/OpenTechBFG/neo/d3xp/MultiplayerGame.cpp:671:13: error: ‘%i’ directive writing between 1 and 8 bytes into a region of size between 3 and 13 [-Werror=format-overflow=]
 const char* idMultiplayerGame::GameTime()
             ^~~~~~~~~~~~~~~~~
.../build/OpenTechBFG/neo/d3xp/MultiplayerGame.cpp:671:13: note: directive argument in the range [-4294940, 4294943]
.../build/OpenTechBFG/neo/d3xp/MultiplayerGame.cpp:721:15: note: ‘sprintf’ output between 5 and 22 bytes into a destination of size 16
   std::sprintf( buff, "%i:%i%i", m, t, s );
   ~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~
cc1plus: some warnings being treated as errors
neo/CMakeFiles/OpenTechEngine.dir/build.make:4502: recipe for target 'neo/CMakeFiles/OpenTechEngine.dir/d3xp/MultiplayerGame.cpp.o' failed
make[2]: *** [neo/CMakeFiles/OpenTechEngine.dir/d3xp/MultiplayerGame.cpp.o] Error 1
CMakeFiles/Makefile2:1479: recipe for target 'neo/CMakeFiles/OpenTechEngine.dir/all' failed
make[1]: *** [neo/CMakeFiles/OpenTechEngine.dir/all] Error 2
Makefile:151: recipe for target 'all' failed
make: *** [all] Error 2
```
